cfme/physical/physical_rack.py

```diff
@@ -0,0 +1,336 @@
+        if ui:
+            self.load_details_page()
+
+        # Verify that the stats retrieved from wrapanapi match those retrieved
+        # from the UI
+        for stat in stats_to_match:
+            try:
+                cfme_stat = int(getattr(self, stat)(method='ui' if ui else None))
+                rack_stat = int(rack_stats[stat])
+
+                if rack_stat != cfme_stat:
+                    msg = "The {} stat does not match. (rack: {}, rack stat: {}, cfme stat: {})"
+                    raise StatsDoNotMatch(msg.format(stat, self.name, rack_stat, cfme_stat))
+            except KeyError:
+                raise RackStatsDoesNotContain(
+                    "Rack stats information does not contain '{}'".format(stat))
+            except AttributeError:
+                raise ProviderHasNoProperty("Provider does not know how to get '{}'".format(stat))
+
+        # Verify that the inventory retrieved from wrapanapi match those retrieved
+        # from the UI
+        for inventory in inventory_to_match:
+            try:
+                cfme_inventory = getattr(self, inventory)(method='ui' if ui else None)
+                rack_inventory = rack_inventory[inventory]
+
+                if rack_inventory != cfme_inventory:
+                    msg = "The {} inventory does not match. (rack: {}, rack inventory: {}, " \
+                          "cfme inventory: {})"
+                    raise StatsDoNotMatch(msg.format(inventory, self.name, rack_inventory,
+                                                     cfme_inventory))
+            except KeyError:
+                raise RackStatsDoesNotContain(
+                    "Rack inventory information does not contain '{}'".format(inventory))
+            except AttributeError:
+                msg = "Provider does not know how to get '{}'"
+                raise ProviderHasNoProperty(msg.format(inventory))
+
+
+@attr.s
+class PhysicalRackCollection(BaseCollection):
+    """Collection object for the :py:class:`cfme.infrastructure.host.PhysicalRack`."""
+
+    ENTITY = PhysicalRack
+
+    def select_entity_rows(self, physical_racks):
+        """ Select all physical rack objects """
+        physical_racks = list(physical_racks)
+        checked_physical_racks = list()
+        view = navigate_to(self, 'All')
+
+        for physical_rack in physical_racks:
+            view.entities.get_entity(name=physical_rack.name, surf_pages=True).check()
+            checked_physical_racks.append(physical_rack)
+        return view
+
+    def all(self):
+        """returning all physical_racks objects"""
+        physical_rack_table = self.appliance.db.client['physical_racks']
+        ems_table = self.appliance.db.client['ext_management_systems']
+        physical_rack_query = (
+            self.appliance.db.client.session
+                .query(physical_rack_table.name, ems_table.name)
+                .join(ems_table, physical_rack_table.ems_id == ems_table.id))
+        provider = None
```

Comment:
slight change here  to make this pattern a bit more concise:

```
provider = self.filters.get('provider')
if provider:
    physical_rack_query.filter(<name match>)
return [
    self.instantiate(name, provider or get_crud_by_name(ems_name))
    for name, ems_name in physical_rack_query.all()
]

Reply:
You can also reduce the calls to `get_crud_by_name` by creating a dictionary of `ems_name: provider_object` key/value pairs from the ems_names that you've found, then looking them up within the for loop to generate entity items.